File: Bronze/2015/December/03_contaminated_milk/ContaminatedMilk.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */
public class ContaminatedMilk {
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("badmilk.in"));
        PrintWriter pw = new PrintWriter("badmilk.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int d = Integer.parseInt(st.nextToken());
        int s = Integer.parseInt(st.nextToken());
        // 记录喝牛奶的信息[人,奶,时间]
        List<List<Integer>> drinkInfo = new ArrayList<>();
        HashSet<Integer> targetMilk = new HashSet<>();
        for (int i = 0; i < d; i++) {
            st = new StringTokenizer(r.readLine());
            List<Integer> info = new ArrayList<>();
            info.add(Integer.parseInt(st.nextToken()));
            info.add(Integer.parseInt(st.nextToken()));
            info.add(Integer.parseInt(st.nextToken()));
            drinkInfo.add(info);
            targetMilk.add(info.get(1));
        }
        // 按人员编号排序 编号相同时按时间排序 为了满足后续双指针实现
        drinkInfo.sort((a, b) -> {
            if (a.get(0).equals(b.get(0))) {
                return a.get(2) - b.get(2);
            }
            return a.get(0) - b.get(0);
        });
        // 记录生病信息[人,时间]
        List<List<Integer>> sickInfo = new ArrayList<>();
        for (int i = 0; i < s; i++) {
            st = new StringTokenizer(r.readLine());
            List<Integer> info = new ArrayList<>();
            info.add(Integer.parseInt(st.nextToken()));
            info.add(Integer.parseInt(st.nextToken()));
            sickInfo.add(info);
        }
        // 按人员编号排序
        sickInfo.sort((a, b) -> a.get(0) - b.get(0));
        // 双指针遍历生病信息的同时 依次缩减坏牛奶的范围
        int j = 0;
        for (int i = 0; i < s; i++) {
            List<Integer> curSickInfo = sickInfo.get(i);
            int curSickPerson = curSickInfo.get(0);
            int curSickTime = curSickInfo.get(1);
            HashSet<Integer> candidateMilk = new HashSet<>();
            while (j < d && drinkInfo.get(j).get(0) < curSickPerson) {
                j++;
            }
            while (j < d && drinkInfo.get(j).get(0) == curSickPerson) {
                if (drinkInfo.get(j).get(2) < curSickTime) {
                    candidateMilk.add(drinkInfo.get(j).get(1));
                }
                j++;
            }
            targetMilk.retainAll(candidateMilk);
        }
        // 找到喝过坏牛奶的人
        HashSet<Integer> sickPerson = new HashSet<>();
        for (int i = 0; i < d; i++) {
            if (targetMilk.contains(drinkInfo.get(i).get(1))) {
                sickPerson.add(drinkInfo.get(i).get(0));
            }
        }
        pw.println(sickPerson.size());
        pw.close();
    }
}

File: Bronze/2015/December/03_contaminated_milk/contaminated_milk.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

int main() {
    freopen("badmilk.in", "r", stdin);
    freopen("badmilk.out", "w", stdout);

    int n, m, d, s;
    cin >> n >> m >> d >> s;
    // 记录喝牛奶的信息[人,奶,时间]
    vector<vector<int>> drink_info;
    unordered_set<int> target_milk;
    for (int i = 0; i < d; i++) {
        vector<int> info(3);
        cin >> info[0] >> info[1] >> info[2];
        drink_info.push_back(info);
        target_milk.insert(info[1]);
    }
    // 按人员编号排序 编号相同时按时间排序 为了满足后续双指针实现
    sort(drink_info.begin(), drink_info.end(), [](const vector<int>& a, const vector<int>& b) {
        if (a[0] == b[0]) {
            return a[2] < b[2];
        }
        return a[0] < b[0];
    });
    // 记录生病信息[人,时间]
    vector<vector<int>> sick_info;
    for (int i = 0; i < s; i++) {
        vector<int> info(2);
        cin >> info[0] >> info[1];
        sick_info.push_back(info);
    }
    // 按人员编号排序
    sort(sick_info.begin(), sick_info.end(), [](const vector<int>& a, const vector<int>& b) {
        return a[0] < b[0];
    });
    // 双指针遍历生病信息的同时 依次缩减坏牛奶的范围
    int j = 0;
    for (int i = 0; i < s; i++) {
        int cur_sick_person = sick_info[i][0];
        int cur_sick_time = sick_info[i][1];
        unordered_set<int> candidate_milk;
        while (j < d && drink_info[j][0] < cur_sick_person) {
            j++;
        }
        while (j < d && drink_info[j][0] == cur_sick_person) {
            if (drink_info[j][2] < cur_sick_time) {
                candidate_milk.insert(drink_info[j][1]);
            }
            j++;
        }
        for (auto it = target_milk.begin(); it != target_milk.end();) {
            if (candidate_milk.find(*it) == candidate_milk.end()) {
                it = target_milk.erase(it);
            } else {
                ++it;
            }
        }
    }
    // 找到喝过坏牛奶的人
    unordered_set<int> sick_person;

    for (int i = 0; i < d; i++) {
        if (target_milk.find(drink_info[i][1]) != target_milk.end()) {
            sick_person.insert(drink_info[i][0]);
        }
    }
    cout << sick_person.size() << endl;

    return 0;
}
